edb/edgeql/codegen.py

```diff
@@ -545,13 +560,26 @@ def visit_Parameter(self, node: qlast.Parameter) -> None:
         self.write(param_to_str(node.name))
 
     def visit_StringConstant(self, node: qlast.StringConstant) -> None:
-        self.write(node.quote, node.value, node.quote)
+        if node.value.isprintable():
+            for d in ("'", '"', '$$'):
+                if d not in node.value:
+                    if '\\' in node.value and d != '$$':
+                        self.write('r', d, node.value, d)
+                    else:
+                        self.write(d, node.value, d)
+                    return
+        self.write(repr(node.value))
```

Comment:
The idea is that we want codegen to be able to output human-friendly code (and at some point it will have different modes of operation: human-friendly, verbose, compact, etc.), which means that certain strings should be raw specifically because it makes it easier to read them. I like the rules for detecting quotes, but I would say that if a string contains a `$$` value we probably should assume that it's likely one raw (dollar) string inside another raw (dollar) string. So instead of potentially rendering it as a `repr` (thus escaping newlines and making it hard to read) we should render it as nested dollar-strings using `dollar_quote_literal` utility to make sure that our outer dollar-quotes aren't matched inside.

Reply:
Didn't know such function exists. But there is a problem with it:

Do we allow non-printable characters in the text? I.e. some ugly characters like backspace or escape? Or should dollar_quote_literal work only with printable characters too?

Reply:
Added into `isprintable` branch but it doesn't seem to change much as `\n` is non-printable.

Reply:
Non-printable characters is a good point. So the first check should be for them, but there should also be a short white-list of non-printable characters that are OK (such as `\r`, `\n`, `\t` if it's for some reason considered non-printable). If the string contains any non-white-listed non-printable characters then it *must* be rendered as an escaped string (via the `repr`). The logic for the rest remains as outlined above.

Reply:
Tweaked a bit in b3459547 but let's not stuck in this loop indefinitely.